engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java

```diff
@@ -9,9 +9,10 @@
 import org.enso.interpreter.service.ExecutionService;
 
 /** A storage for computed values. */
-public final class RuntimeCache {
+public final class RuntimeCache implements java.util.function.Function<String, Object> {
 
   private final Map<UUID, SoftReference<Object>> cache = new HashMap<>();
+  private final Map<UUID, SoftReference<Object>> all = new HashMap<>();
```

Comment:
Values are disappearing from `cache` too much. I had to add `all` that's never cleared to increase the chance the argument values are available when widget annotations are computed.

Reply:
Caching works this way. The `CachePreferenceAnalysis` assigns weights to the expressions to mark them for caching. Currently, it only marks the righthand side of assignment expressions, and only those expressions end up being cached.

If we start storing intermediate expressions, like the function arguments, between program executions, we may end up consuming all available memory affecting the performance. One of the ways how to keep the heap size from growing uncontrollably that we discussed before is [Poor Richard's Memory Manager](https://cse.buffalo.edu/~mhertz/prmm-ismm-2011.pdf).

Another issue is that when we start caching intermediate results, we can't guarantee they won't be evicted from the cache under the memory pressure. There should be a way to recover from the cache misses, i.e. re-computing a particular expression

Reply:
@4e6, what is the **lifecycle of a `RuntimeCache`**? When it gets created? When it is removed? 

Why there isn't a [single cache per IdExecutionInstrument](https://github.com/enso-org/enso/pull/9410#discussion_r1567589821) being around while the engine+LS session is running? And yes, I am reading [caching.md](https://github.com/enso-org/enso/blob/fda41cbfd1d1966d8772d1a889905b78651e734e/docs/runtime/caching.md).

Reply:
I believe I managed to solve this problem in 29e788c!

After a discussion with @4e6 and @hubertp today we concluded that the most common case in the IDE is currently
```ruby
operator1.join operator2
```
e.g. the `right` argument expression is a direct reference to LHS of a _previous assignment_. As such the value of the `right` argument expression is already in the `RuntimeCache`! It is just not associated with the _`right` argument expression UUID_.  29e788c fixes that
- it keeps mapping from values to UUIDs that introduced them (using _weak references_ - e.g. not preventing GC)
- whenever a value that is already in the `RuntimeCache` shall be **cached again** under different UUID, it is inserted into the cache via a `WeakReference` (e.g. not preventing GC)

The whole system continues to work and we cannot be _wasting more memory_ than we use now. I believe this is  a [good enough](http://wiki.apidesign.org/wiki/Good) compromise for now.